Cython/Compiler/ExprNodes.py

```diff
@@ -9371,21 +9371,22 @@ def analyse_default_args(self, env):
         # so their optional arguments must be static, too.
         # TODO: change CyFunction implementation to pass both function object and owning object for method calls
         must_use_constants = env.is_c_class_scope or (self.def_node.is_wrapper and env.is_module_scope)
-
         for arg in self.def_node.args:
-            if arg.default and not must_use_constants:
-                if not arg.default.is_literal:
-                    arg.is_dynamic = True
-                    if arg.type.is_pyobject:
-                        nonliteral_objects.append(arg)
+            if arg.default:
+                if not must_use_constants:
+                    if not arg.default.is_literal:
+                        arg.is_dynamic = True
+                        if arg.type.is_pyobject:
+                            nonliteral_objects.append(arg)
+                        else:
+                            nonliteral_other.append(arg)
                     else:
-                        nonliteral_other.append(arg)
-                else:
-                    arg.default = DefaultLiteralArgNode(arg.pos, arg.default)
-                if arg.kw_only:
-                    default_kwargs.append(arg)
-                else:
-                    default_args.append(arg)
+                        arg.default = DefaultLiteralArgNode(arg.pos, arg.default)
+                if arg.type.is_pyobject or arg.type.is_numeric:
```

Comment:
```suggestion
                if arg.default.type and arg.default.type.can_coerce_to_pyobject(env):
```

This seems to pass `python runtests.py cyfunction default signature` (i.e. all the tests in this question, plus some related ones)